docs/en/stores/supplemental.md

```diff
@@ -0,0 +1,425 @@
+], [ undoOnFailure ])
+```
+
+If any of the commands fail during their execution the `undoOnFailure` middleware will have an opportunity to apply `undoOperations`.
+
+It is important to note that `undoOperations` only apply to the commands fully executed during the process. It will not contain any operations to rollback state that changed as a result of other processes that may be executed asynchronously or state changes performed in middleware or directly on the store itself. These use cases are outside the scope of the undo system.
+
+## Optimistic Updates
+
+Optimistic updating can be used to build a responsive UI despite interactions that might take some time to respond, for example saving to a remote resource.
+
+In the case of adding a todo item for instance, with optimistic updating, we can immediately add the todo before we even make a request to the server and avoid having an unnatural waiting period or loading indicator. When the server responds we can then reconcile the outcome based on whether it is successful or not.
+
+In the success scenario, we might need to update the added Todo item with an id that was provided in the response from the server and change the color of the Todo item to green to indicate it was successfully saved.
+
+In the error scenario, it might be that we want to show a notification to say the request failed and turn the Todo item red, with a "retry" button. It's even possible to revert/undo the adding of the Todo item or anything else that happened in the process.
+
+> Undo example
+
+```ts
+const handleAddTodoErrorProcess = createProcess('error', [ () => [ add(path('failed'), true) ]; ]);
+
+const addTodoErrorMiddleware = () => {
+	return {
+		after: () => (error, result) {
+			if (error) {
+				result.store.apply(result.undoOperations);
+				result.executor(handleAddTodoErrorProcess);
+			}
+		}
+	};
+};
+
+const addTodoProcess = createProcess('add-todo', [
+		addTodoCommand,
+		calculateCountsCommand,
+		postTodoCommand,
+		calculateCountsCommand
+	],
+	[ addTodoCallback ]);
+```
+
+-   `addTodoCommand` adds the new todo into the application state
+-   `calculateCountsCommand` recalculates the count of completed and active todo items
+-   `postTodoCommand` posts the todo item to a remote service and using the process `after` middleware we can make changes if there is a failure
+    -   on _failure_ the changes are reverted and the failed state field is set to true
+    -   on _success_ update the todo item id field with the value received from the remote service
+-   `calculateCountsCommand` runs again after the success of `postTodoCommand`
+
+### Synchronized updates
+
+In some cases it is better to wait for the server to update for the process to wait until a remote call has been completed. For instance when a process will remove an element from the screen or when the outlet changes to display a different view. Restoring a state that triggered these type of actions can be surprising.
```

Comment:
"In some cases it is better to wait for the server to update for the process to wait until a remote call has been completed."

I'm having trouble following this.

Reply:
"For instance when a process will remove an element from the screen or when the outlet changes to display a different view. Restoring a state that triggered these type of actions can be surprising." -> "For example, when a process will remove an element from the screen or when the outlet changes to display a different view, restoring a state that triggered these actions can be surprising."